### PR TITLE
[SPARK-9003] [MLlib] Add mapActive{Pairs,Values}, map{Pairs,Values} for MLlib/Vector

### PR DESCRIPTION
**Pull request by yanboliang**
Add mapActive{Pairs,Values}, map{Pairs,Values} function to MLlib/Vector refer the Breeze's API. 
MLlib Vector/Matrix is not supposed to become yet another vector/matrix libaray. But map function is important enough to become the interface of mllib/vector just like foreachActive which is supported at present. And MLlib/Matrix has already supported map/update/foreachActive function.

### CODE REVIEW

**Comment by yanboliang** [comment]
@mengxr

**Comment by feynmanliang** [comment]
Perhaps we should consider following [Breeze's API](http://www.scalanlp.org/api/breeze/index.html#breeze.linalg.DenseVector) here and provide:

 * `mapPairs` / `mapValues` : the former includes the indices as arguments (i.e. `f : (Int, Double) => Double`) while the latter is what you have now
 * `mapActive{Pairs,Values}` / `map{Pairs,Values}` : the former only iterates over active elements (what you have now) while the latter iterates over all elements

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 1]
```diff
@@ -692,6 +744,29 @@ class SparseVector @Since("1.0.0") (
 
   private[spark] override def toBreeze: BV[Double] = new BSV[Double](indices, values, size)
 
+  private[spark] override def mapActiveValues(f: Double => Double): Vector =
+    new SparseVector(size, indices.clone(), values.map(f))
+
+  private[spark] def mapActivePairs(f: (Int, Double) => Double): Vector = {
```
Comment (Mageswaran1989) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: I am learning Scala. Could you please explain how this maps only the active pairs? 
what is happening in the line no: 756

**Comment by feynmanliang** [comment]
@yanboliang I think there are some [discussions about whether we should enrich the Vectors API](https://www.mail-archive.com/user@spark.apache.org/msg35434.html)

CC @srowen

**Comment by srowen** [comment]
Yeah, I think the question is, what's the need / motivation? I don't see this being used in the code. If it's just because it sounds useful, I'm not sure that's enough justification. If it can simplify or speed up project code, that's an argument worth advancing.

**Pull request closed without merging**